contrib/parseq-batching/src/main/java/com/linkedin/parseq/batching/BatchingStrategy.java

```diff
@@ -0,0 +1,158 @@
+    return Task.async(desc, ctx -> {
+      final SettablePromise<T> result = Promises.settable();
+      Long planId = ctx.getPlanId();
+      BatchBuilder<K, T> builder = _batches.get(planId);
+      if (builder == null) {
+        builder = Batch.builder();
+        BatchBuilder<K, T> existingBuilder = _batches.putIfAbsent(planId, builder);
+        if (existingBuilder != null) {
+          builder = existingBuilder;
+        }
+      }
+      builder.add(key, ctx.getShallowTraceBuilder(), result);
+      return result;
+    });
+  }
+
+  private Task<?> taskForBatch(final G group, final Batch<K, T> batch) {
+    return Task.async(group.getName(batch), ctx -> {
+      final SettablePromise<T> result = Promises.settable();
+      final PromiseListener<T> countDownListener =
+          new CountDownPromiseListener<T>(batch.size(), result, null);
+
+      //one of the batchable tasks becomes a parent of batch task
+      //all other tasks become potential parents
+      final TraceBuilder traceBuilder = ctx.getTraceBuilder();
+      Relationship rel = Relationship.PARENT_OF;
+      for (BatchEntry<T> entry : batch.values()) {
+        for (ShallowTraceBuilder shallowTraceBuilder: entry.getShallowTraceBuilders()) {
+          traceBuilder.addRelationship(rel, shallowTraceBuilder, ctx.getShallowTraceBuilder());
+          rel = Relationship.POTENTIAL_PARENT_OF;
+        }
+        entry.getPromise().addListener(countDownListener);
+      }
+
+      try {
+        if (batch.size() == 1) {
+          Entry<K, BatchEntry<T>> entry = batch.entires().iterator().next();
+          executeSingleton(group, entry.getKey(), entry.getValue());
+        } else {
+          executeBatch(group, batch);
+        }
+      } catch (Throwable t) {
+        batch.failAllRemaining(t);
+      }
+
+      return result;
+    });
+  }
+
+  private Collection<Task<?>> taskForBatches(Collection<Entry<G, Batch<K, T>>> batches) {
+    return batches.stream().map(entry -> taskForBatch(entry.getKey(), entry.getValue())).collect(Collectors.toList());
+  }
+
+  void handleBatch(final PlanContext planContext) {
+    final BatchBuilder<K, T> batchBuilder = _batches.remove(planContext.getId());
+    if (batchBuilder != null) {
+      final Batch<K, T> batch = batchBuilder.build();
+      if (batch.size() > 0) {
+        try {
+          final Map<G, Batch<K, T>> batches = split(batch);
+          if (batches.size() > 0) {
+            taskForBatches(batches.entrySet()).forEach(task -> new ContextImpl(planContext, task).runTask());
+          }
+        } catch (Throwable t) {
+          //we don't care if some of promises have already been completed
+          //all we care is that all remaining promises have been failed
+          batch.failAllRemaining(t);
+        }
+      }
+    }
+  }
+
+  public abstract void executeBatch(G group, Batch<K, T> batch);
+
+  public abstract void executeSingleton(G group, K key, BatchEntry<T> entry);
```

Comment:
I'm a little concerned about this API. It seems BatchEntry is more like a class for parseq internal use, but here we expose it to user code. This leaks too much (e.g. addShallowTraceBuilder, getShallowTraceBuilder). 

Can we use use SettablePromise here?